src/util/resourceValidationUtils.js

```diff
@@ -0,0 +1,124 @@
+const { resolveSchema } = require('@projecttacoma/node-fhir-server-core');
+const axios = require('axios');
+const { v4: uuidv4 } = require('uuid');
+
+/**
+ * Checks body of request by querying a FHIR validation server. Updates the submitted resource to
+ * include validated profiles in the meta.profile field. Responds with 400 if body is not valid FHIR
+ * 500 if validation server error
+ * @param {Object} req request object of an express request
+ * @param {Object} res response object of an express request
+ * @param {Function} next express function which passes info along after clearing the middleware
+ */
+async function validateFhir(req, res, next) {
+  const profiles = retrieveProfiles(req.originalUrl, req.body);
+  if (profiles !== 'Calculation') {
+    //We know the profile pulled from the URL will be the last profile added to the array
+    const qicoreProfile = `http://hl7.org/fhir/us/qicore/StructureDefinition/qicore-${profiles[
+      profiles.length - 1
+    ].toLowerCase()}`;
+    console.log(profiles);
+    const qicoreValidationInfo = await getValidationInfo([qicoreProfile], req.body, req.base_version);
+    const validationInfo = await getValidationInfo(profiles, req.body, req.base_version);
+
+    if (validationInfo.isValid) {
+      /*
+       * We don't want the qicore failure to cause the whole resource to fail validation,
+       * so simply don't add it and move on if it fails
+       */
+      if (qicoreValidationInfo.isValid) {
+        profiles.push(qicoreProfile);
+      }
+      if (req.body?.meta?.profile) {
+        const validatedProfiles = new Set(req.body.meta.profile);
+        profiles.forEach(profile => {
+          validatedProfiles.add(profile);
+        });
+        req.body.meta.profile = Array.from(validatedProfiles);
+      } else {
+        req.body['meta'] = { profile: profiles };
+      }
+
+      next();
+    } else {
+      res.status(validationInfo.code).json(validationInfo.data);
+    }
+  } else {
+    next();
+  }
+}
+
+/**
+ * Determines the profiles the submitted resource is expected to be validated against
+ * @param {String} originalUrl the url the request was sent to
+ * @param {Object} body an express request body containing a FHIR resource
+ * @returns String or Array
+ */
+function retrieveProfiles(originalUrl, body) {
+  const profiles = [];
+  const params = originalUrl.split('/');
+
+  const metaProfiles = body?.meta?.profile;
+  if (metaProfiles) {
+    profiles.push(...metaProfiles);
+  }
+  if (params[params.length - 1] === '$submit-data') {
+    profiles.push('Parameters');
+  }
+  // We don't need to validate posted Parameters bodies for dollar-sign operations since these aren't stored in the db
+  else if (params[params.length - 1][0] === '$') {
+    return 'Calculation';
```

Comment:
I see that we use this `'Calculation'` string to indicate when we don't care about doing validation (the whole validation logic above is wrapped in an `if (profiles !== 'Calculation')`. This confused me a bit at first and I had to trace it down to this function.

I think it would be a little better to just return `[]` here, as that makes more conceptual sense for this function: you ask it to retrieve the profiles you want to validate against, and it certain cases you give it nothing back. Then, the above if statement can just be `if (profiles.length > 0)`